### DISCUSSION TITLE
markdownRendererInputFencedCode + minted package

### DISCUSSION THREAD

**Discussion by csrmy** [Witiko/markdown#485] [Q&A] [answered]
Hi, thanks in advance for your help.

I try to style and add code highlight to the fenced blocks of code from markdown to LaTex.

I successfully setup an override for the `\markdownRendererInputFencedCode` and the `\markdownRendererInputVerbatim` to be able to draw a box around the rendering of the fenced blocks of code. I have trouble to combine my implementation with the [`minted`](https://ctan.org/pkg/minted) package.

I tried to use `\begin{minted}` and `\inputminted` but I have errors everytime when adding it in the override of `\markdownRendererInputFencedCode`.

Do you have any idea how I could achieve that ?

tex file:
```
\documentclass{article}

\usepackage{markdown}
\usepackage{tcolorbox}

% markdown settings
\markdownSetup{%
	fencedCode=true,
	fancyLists=true,
	texMathDollars=true,
	inlineNotes=true,
	notes=true,	
	pipeTables=true,
	taskLists=true,
	underscores=false,
	subscripts=true,
	superscripts=true	
}


% Define code block style
\newtcolorbox{codeblock}[2][]{
title={#2},
fonttitle=\scriptsize,
leftrule=1mm,
rightrule=0mm,
toprule=0mm,
bottomrule=0mm,
arc=0mm,
#1}

\global\def\markdownRendererInputVerbatim#1{%
    \verbatiminput{#1}%
}

\global\def\markdownRendererInputFencedCode#1#2#3{%
\begin{codeblock}{#2}%
\markdownRendererInputVerbatim{#1}%
\end{codeblock}
}

\begin{document}
\markdownInput{test.md}
\end{document}
```

test.md file:

``` python
print("hello")
```

``` c
int main(int ac, char** av)
{
    return 0;
}
```
~~~ js
if (a > 3) {
    moveShip(5 * gravity, DOWN);
}
~~~~~~

Result:
![image](https://github.com/user-attachments/assets/2e3b5e69-63f1-4073-9918-0c1329e46eba)

**Comment by Witiko** [comment]
> I tried to use `\begin{minted}` and `\inputminted` but I have errors everytime when adding it in the override of `\markdownRendererInputFencedCode`.

In your definition of `\markdownRendererInputFencedCode`, you should replace `\markdownRendererInputVerbatim{#1}` with `\inputminted{#2}{#1}`. This will render file `#1` using [a Pygments lexer](https://pygments.org/docs/lexers/) for language `#2`, such as `python`, `c`, and `js`.

You must always specify a lexer, so that `minted` knows how to highlight the syntax. See also [the documentation of `minted`](http://mirrors.ctan.org/macros/latex/contrib/minted/minted.pdf).

**Reply by csrmy** [reply]
Thank you for your reply.

Yes that is something that I tried already while reading the `minted` package documentation.

If I change this in the tex file:
```
\global\def\markdownRendererInputVerbatim#1{%
    \verbatiminput{#1}%
}

\global\def\markdownRendererInputFencedCode#1#2#3{%
\begin{codeblock}{#2}%
	\inputminted{#2}{#1}
\end{codeblock}
}
```

Error message:
```
! Use of \FV@PYG doesn't match its definition.
\PYG #1#2->\FV@PYG {#1}{\FancyVerbBreakStart #2\FancyVerbBreakStop }
l.2 ...YG{l+s+s2}{\PYGZdq{}hello\PYGZdq{}}\PYG{p}{)}
If you say, e.g., `\def\a1{...}', then you must always
put `1' after `\a', since control sequence names are
made up of letters only. The macro here has not been
followed by the required stuff, so I'm ignoring it.
```

I thought that I was doing something wrong logic-wise so I asked for help.
Thanks again for your time !

**Reply by Witiko** [reply]
It may have something to do with your environment `codeblock`. Does the code work when you remove `\begin{codeblock}{#2}` and `\end{codeblock}` from your definition of `\markdownRendererInputFencedCode`?

**Reply by csrmy** [reply]
No difference with:
```
\global\def\markdownRendererInputFencedCode#1#2#3{%
\inputminted{#2}{#1}%
}
```
***
If I do:
```
\begin{document}
\inputminted{c}{test.md}
\end{document}
```
test.md:
```
int main(int ac, char** av)
{
    if (a > 3)
    {
        moveShip(5 * gravity, DOWN);
    }
    return 0;
}
```
Result:
![image](https://github.com/user-attachments/assets/457106cd-af3d-4bc4-aa21-e457f9e2b028)

So I thought It have to do with the usage of the `markdown` package since it's while using renderer override that I get those issues.

**Reply by csrmy** [reply]
@Witiko I allow myself to run this topic again since I could not find a solution by myself despite looking into it the past days.
Do you have any idea of what I can do about it since it's not due to the `codeblock` environement I was using ?

**Reply by Witiko** [reply]
What output do you get when you use the following definition, I wonder:

``` tex
\global\def\markdownRendererInputFencedCode#1#2#3{%
  \texttt{\textbackslash inputminted\{#2\}\{#1\}}%
}
```

**Reply by csrmy** [reply]
![image](https://github.com/user-attachments/assets/f601d8be-12d6-4ca1-9f72-0c1d80e97d94)

**Reply by Witiko** [reply]
This seems as expected. I assume that, if you wrote `\inputminted{c}{./_markdown_test/8a1c...1ece.verbatim}` (replace `...` with the rest of the digits) directly in your document, the source code would by correctly typeset by `minted`.

The issue seems to be caused by the fact that we use a special catcode regime inside `\begin{markdown} ... \end{markdown}` in which `#` and `%` are treated as ordinary characters without any special meaning. While this makes it easier to work with the result of the conversion, especially when the `hybrid` option is enabled, it may break packages such as minted.

The solution, which is also used by the default fenced code renderer, is to reset the catcode of `#` before using minted, as follows:

https://github.com/Witiko/markdown/blob/3cf013f2340eaa50177bbea2c376c208f4ba6031/markdown.dtx#L36616-L36621

In your specific case, you would use the following definition:

``` tex
\global\def\markdownRendererInputFencedCode#1#2#3{%
  \catcode`\#=6\relax
  \begin{codeblock}{#2}%
    \inputminted{#2}{#1}%
  \end{codeblock}%
  \catcode`\#=12\relax
}
```

I am sorry that I did not notice earlier.

**Reply by csrmy** [reply]
Hi @Witiko, thank you for your answer.
It works on my side perfectly !

But I am struggling setting up a conditional logic with `inputminted` and `markdownRendererInputVerbatim`.
Since the language must be provided for the minted package to work, I have to check if the fenced block title is empty.

Thats what I did:
```
\global\def\markdownRendererInputFencedCode#1#2#3{%
\catcode`\#=6\relax
\begin{codeblock}{#2}%
	\if #2\empty
		\markdownRendererInputVerbatim{#1}
	\else
		\inputminted{#2}{#1}%
	\fi
\end{codeblock}
\catcode`\#=12\relax
}
```

But it doesn't work when there is not title, it works when there is one though. And I never did such thing in Tex, I tried to search on the web but cannot implement anything that works, do you have an idea ?

**Reply by Witiko** [reply]
You may wish to have a look at how the default `inputFencedCode` renderer is implemented:

https://github.com/Witiko/markdown/blob/3cf013f2340eaa50177bbea2c376c208f4ba6031/markdown.dtx#L36575-L36648

We use [the LaTeX3 programming language](https://ctan.org/pkg/l3kernel?lang=en) between `\ExplSyntaxOn` and `\ExplSyntaxOff`, which makes programming easier.

**Reply by csrmy** [reply]
@Witiko Thank you so much for your time you are awesome !

**Comment by csrmy** [comment, marked as answer]
# Answer
After @Witiko 's help.
Now the renderer override `\markdownRendererInputFencedCode` will use the minted package only if the fenced block "title" is present, using it as a language for the [`minted`](https://ctan.org/pkg/minted) package while using our own environment to wrap the style of the fenced code bloc.

main.tex:
``` tex
% Create a new environement "codeblock"
\newtcolorbox{codeblock}[2][]{
title={#2},
fonttitle=\scriptsize,
leftrule=1mm,
rightrule=0mm,
toprule=0mm,
bottomrule=0mm,
arc=0mm,
#1}

% Override the renderer of fenced code blocs
\ExplSyntaxOn% enable LaTex3 syntax
\global\def\markdownRendererInputFencedCode#1#2#3{%
	\catcode`\#=6\relax
	\begin{codeblock}{ #2 }
		\tl_if_empty:nTF
		{ #2 }% if the second argument (fence block "title" is empty)
		{ \markdownRendererInputVerbatim{#1} }% render regular verbatim
		{ \inputminted{#2}{#1} }% else render minted text
	\end{codeblock}
	\catcode`\#=12\relax
}
\ExplSyntaxOff% disable LaTex3 syntax
```
> `codeblock` is a custom environment created with `\newtcolorbox` from the [`tcolorbox`](https://ctan.org/pkg/tcolorbox) package. 

## Result
test-language.md:
~~~
``` c++
int main(int arc, char** arv)
{
    std::cout << "Hello World !" << std::endl;
    return 0;
}
```
~~~
![image](https://github.com/user-attachments/assets/4ede66f2-cac1-4938-96d8-55c50be280a8)
***
test-no-language.md:
~~~
```
int main(int arc, char** arv)
{
    std::cout << "Hello World !" << std::endl;
    return 0;
}
```
~~~
![image](https://github.com/user-attachments/assets/cb7e3a08-97a7-46a0-81df-8b2c67647e40)
